### PR TITLE
Make it possible to use Stack logging in Docker

### LINKED ISSUE #62758: Include the default log4j config in Docker images
At present we ship a custom log4j configuration in our Docker images. This done to make the images more "Docker-like", since the convention with Docker is to write all container output to stdout and stderr.

This is not ideal, because:

   * It is more difficult for Filebeat to use the right `event.dataset` value
   * There can be drift between the configurations
   * Users can't easily choose to use the default logging config and have to bind-mount their own config
   * Writing to stdout / stderr doesn't extend to the logging configurations that plugins can supply

Elasticsearch should make it easy to pick what style of logging to use in Docker. Ideally, the stdout/stderr logging config should be automatically derived from the default configs, so as to minimise drift.

### PR DESCRIPTION
**Pull request by pugnascotia**
Closes #62758.

Include the Stack log4j config in the Docker image, in order to make it
possible to write logs in a container environment in the same way as for
an archive or package deployment. This is useful in situations where the
user is bind-mounting the logs directory and has their own arrangements
for log shipping.

To use stack logging, set the environment variable `ES_LOG_STYLE` to
`file`. It can also be set to `console`, which is the same as not
specifying it at all.

The Docker logging config is now auto-generated at image build time, by
running the default config through a transformer program when preparing
the distribution in an image builder step.

Implementation notes:

   * Regarding the docker config change -  The major constraint on the implementation here is the need to publish reproducible build files to Docker hub. It means that all you have to work with is the archive distro of ES and whatever you put in the build context. For that reason I opted for an executable jar, since the v8.0 image has almost nothing useful in it otherwise. Same thing goes for Iron Bank - they build the images, not us.
   * In the docker distribution `build.gradle`, I changed a helper closure into a class with a static method in order to work around an issue where the Docker image was always being rebuilt, even when there were no changes.

### CODE REVIEW

File: distribution/docker/src/docker/bin/docker-entrypoint.sh [hunk 1]
```diff
@@ -57,6 +57,21 @@ if [[ -f bin/elasticsearch-users ]]; then
   fi
 fi
 
+if [[ -n "$ES_LOG_STYLE" ]]; then
+  case "$ES_LOG_STYLE" in
+    docker)
+      # This is the default. Nothing to do.
+      ;;
+    stack)
```
Comment (mark-vieira) [inline distribution/docker/src/docker/bin/docker-entrypoint.sh:65]: I'm not in love with this name. It doesn't really mean anything from a user perspective. I'd be more inclined to call this `file` or `disk`.

**Review by mark-vieira — COMMENTED** [review]
This doesn't seem to address the configuration drift issue as described in #62758. So we still have the problem of multiple logging config files that need to be kept in since. If we still intend to address that we should open a separate issue for that.

Second, and mostly from curiosity, is there a specific use case for using the file-based logging? Is this something cloud requested?

**Review by mark-vieira — APPROVED** [review]

File: distribution/docker/src/docker/bin/docker-entrypoint.sh (hunk 1 above)
Comment (pugnascotia) [reply distribution/docker/src/docker/bin/docker-entrypoint.sh]: Agreed, I've swapped to `console` and `file.

**Comment by pugnascotia** [comment]
@mark-vieira yes, Cloud don't want to write to the console since they use filebeat to ship their logs, and would prefer to be able to use the default logging config.

I also make an attempt at transforming the default config into a console-only config. The major constraint on the implementation here is the need to publish reproducible build files to Docker hub. It means that all you have to work with is the archive distro of ES and whatever you put in the build context. For that reason I opted for a jshell script, since the v8.0 image has almost nothing useful in it otherwise. We could probably change it to a standard Java class if you think that using jshell is too gonzo, but we'd still need to put the source in the Docker context in order to provide reproducible builds. Same thing goes for Iron Bank - they build the images, not us.

**Comment by pugnascotia** [comment]
BTW I ran the Docker test for the default and UBI distros locally, and also build the Iron Bank image.

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/2 - known issue.

**Comment by mark-vieira** [comment]
> We could probably change it to a standard Java class if you think that using jshell is too gonzo

My main concern is lack of test coverage. I say let's just punt on that for now. We don't have to solve the config drift issue as part of this PR. My preference however would be to use some kind of log4j2 construct here. A custom "tee" appender perhaps that could send logging to one of two locations based on some config/env var/etc.

**Comment by pugnascotia** [comment]
Looking at the [RoutingAppender](https://logging.apache.org/log4j/2.x/manual/appenders.html#RoutingAppender) that you pointed me at, I believe we'd need to configure it for every appender that we currently, which would be pretty ugly. I'd rather make the transformer script testable.

What about moving it to `distribution/docker/src/main/java`, along with tests, then compiling and running it in the Docker build?

**Comment by mark-vieira** [comment]
> What about moving it to distribution/docker/src/main/java, along with tests, then compiling and running it in the Docker build?

That's probably overkill. Could we put it in a separate subproject, package it in a jar and then ship that jar in the dockerfile. Effectively treat it like one of our cli tools?

**Comment by pugnascotia** [comment]
@mark-vieira OK, we're now using an executable jar instead of jshell.

@breskeby any thoughts on improving the new sub-project?

**Comment by pugnascotia** [comment]
@breskeby I'd particularly like a review on the new gradle config in this PR.

File: distribution/docker/build.gradle [hunk 2]
```diff
@@ -21,13 +21,33 @@ configurations {
...
   aarch64DockerSource project(path: ":distribution:archives:linux-aarch64-tar", configuration:"default")
   dockerSource project(path: ":distribution:archives:linux-tar", configuration:"default")
   aarch64OssDockerSource project(path: ":distribution:archives:oss-linux-aarch64-tar", configuration:"default")
   ossDockerSource project(path: ":distribution:archives:oss-linux-tar", configuration:"default")
+  transformLog4jJar project(path: ":distribution:docker:transform-log4j-config", configuration: "default")
+}
+
+/**
+ * Used in the Dockerfile template to wrap commands in a retry loop. It can't be
+ * a closure because Gradle can't effectively cache those.
```
Comment (breskeby) [inline distribution/docker/build.gradle:37]: Can you elaborate what can't be cached here?

File: distribution/docker/build.gradle [hunk 3]
```diff
@@ -21,13 +21,33 @@ configurations {
...
+ */
+class Retry {
```
Comment (breskeby) [inline distribution/docker/build.gradle:39]: In general I'm in favour to have class definitions living in buildSrc somewhere no matter how small they are. To keep our build scripts _a bit_ cleaner and smaller. Also these classes should change less often than the script itself.

File: distribution/docker/build.gradle [hunk 4]
```diff
@@ -100,7 +108,7 @@ ${indent}${exitKeyword} \$exit_code"""
     'docker_base'         : base.name().toLowerCase(),
     'version'             : VersionProperties.elasticsearch,
     'major_minor_version' : "${major}.${minor}",
-    'retry_loop'          : retry_loop
+    'retry'               : Retry,
```
Comment (breskeby) [inline distribution/docker/build.gradle:111]: I think the , is too much here

**Review by breskeby — APPROVED** [review]
lgtm. some minor remarks. 
 
 I remember there were memory leaking issues reported with using Groovy template expansion as we rely on here in our docker file setups atm. Therefore in the long term ideally we would move away from google template expansion (but totally out of scope here, just as a general comment as we also havn't received related issues reported for the es build afaik).

**Comment by pugnascotia** [comment]
@breskeby I moved the retry class to `buildSrc`.

The logic for build the the `Dockerfile` is already pretty hairy, so I wouldn't be surprised if we choose to spin that off into its own class at some point. It seems to be working well enough at the moment however.

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Pull request merged**